### ISSUE TITLE
Params containing a semi-colon are truncated

### ISSUE THREAD

**Issue by jimkinsey** [scalatra/scalatra#473]
While evaluating 2.4 for an app currently on 2.2.1, we found a number of our routes either fail to match or get incorrect values in parameters.

We realised this happens in the case where the parameter value contains a URL-encoded semi-colon (%3B). When it does, the value is truncated from the point where the semi-colon is located. The test below reproduces this:

```
import org.scalatest.FunSuite
import org.scalatra.ScalatraServlet
import org.scalatra.test.scalatest.ScalatraSuite

class RouteBugTest extends FunSuite with ScalatraSuite {

  val servlet = new ScalatraServlet {

    get("/captures/(.*)".r) {
      multiParams("captures").head
    }

    get("/sinatra/:thing") {
      params("thing")
    }

  }

  addServlet(servlet, "/*")

  test("captures") {
    get("/captures/ok%3Bgo") {
      body should be("ok;go")
    }
  }

  test("sinatra-like params") {
    get("/sinatra/wink%3B)") {
      body should be("wink;)")
    }
  }

}
```

Additionally, if a value contains a semi-colon then it may result in a route-match failure:

```
import org.scalatest.FunSuite
import org.scalatra.ScalatraServlet
import org.scalatra.test.scalatest.ScalatraSuite

class RouteBugTest extends FunSuite with ScalatraSuite {

  val servlet = new ScalatraServlet {

    get("/:thing/length") {
      params("thing").length.toString
    }

  }
  addServlet(servlet, "/*")

  test("route match failure") {
    get("/semi%3Bcolon/length") {
      body should be("10")
    }
  }

}
```

**Comment by eudactus** [comment]
This behaviour appears to start from 2.3.0. Is it a bug or feature that encoded semicolons are interpreted like this? It certainly seems like a bug from my perspective.

The closest test I can find in the project is RouteTest: "should chop off uri starting from semicolon"

**Comment by casualjim** [comment]
ok
We have to decode the path to be able to match it.
';' is a reserved char in HTTP, so it's not a bug afaik.

**Comment by jimkinsey** [comment]
The semi-colon might be a reserved char in HTTP, but an encoded semi-colon is not; that is the point of encoding values in URLs, and I would argue that any truncation should happen *before* decoding rather than after.

IETF RFC-1738:

    If the character corresponding to an octet is reserved in a scheme, the octet must be encoded.  The characters ";", "/", "?", ":", "@", "=" and "&" are the characters which may be reserved for special meaning within a scheme. No other characters may be reserved within a scheme.

Not allowing encoded reserved chars in URLs is practically diverging from the URL standard.

The standard identifiers used in our domain (DOIs) are very flexible in the characters allowed in them and a number of ours do contain semi-colons among other reserved characters:

http://en.wikipedia.org/wiki/Digital_object_identifier

An example containing a semi-colon:

http://link.springer.com/article/10.1672%2F0277-5212%282000%29020%3C0470%3ASDOAFA%3E2.0.CO%3B2

If this is not considered a bug, we will at some point simply have to stop using Scalatra as our web app framework of choice, which I personally feel would be a shame.

**Comment by casualjim** [comment]
how do you propose we support both route definitions like "/bläh" and this use case of semi-colons?

We already do 2 step decoding so I imagine something can be done there.

**Comment by jimkinsey** [comment]
Without being familiar with the code it seems to me that  a rule like "URLs should not contain character x" should only apply to the incoming URL with encoded params, and once the path has been decoded for route matching any characters should be considered valid.

I would expect this to then work with the case you mention - if the a-with-umlaut is encoded in the request it will be decoded for route matching; if not then there is nothing to decode and it would also match.

We are happy to look further into it (@eudactus has already put some work in) as we are quite dependent on Scalatra supporting this, but we need to confirm whether it is an intentional feature.

**Comment by casualjim** [comment]
some more info: http://stackoverflow.com/a/2163885

the code to change is: https://github.com/scalatra/scalatra/blob/2.4.x/core/src/main/scala/org/scalatra/UriDecoder.scala#L13-L17

**Comment by takezoe** [comment]
Fixed in #691
